fe/fe-core/src/main/java/com/starrocks/sql/common/RangePartitionDiffer.java

```diff
@@ -0,0 +1,444 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.starrocks.sql.common;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Predicates;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Range;
+import com.starrocks.analysis.Expr;
+import com.starrocks.catalog.Column;
+import com.starrocks.catalog.Database;
+import com.starrocks.catalog.MaterializedView;
+import com.starrocks.catalog.PartitionInfo;
+import com.starrocks.catalog.PartitionKey;
+import com.starrocks.catalog.Table;
+import com.starrocks.catalog.Type;
+import com.starrocks.common.AnalysisException;
+import com.starrocks.common.Pair;
+import com.starrocks.common.UserException;
+import com.starrocks.common.util.RangeUtils;
+import com.starrocks.common.util.concurrent.lock.LockType;
+import com.starrocks.common.util.concurrent.lock.Locker;
+import com.starrocks.connector.PartitionUtil;
+import com.starrocks.sql.analyzer.SemanticException;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.threeten.extra.PeriodDuration;
+
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import static com.starrocks.sql.common.PRangeCellPlus.toPRangeCellPlus;
+
+// TODO: refactor all related code into this class
+
+/**
+ * Compute the Materialized View partition mapping from base table
+ * e.g. the MV is PARTITION BY date_trunc('month', dt), and the base table is daily partition, the resulted mv
+ * should be monthly partition
+ */
+public final class RangePartitionDiffer extends PartitionDiffer {
+    private static final Logger LOG = LogManager.getLogger(RangePartitionDiffer.class);
+
+    private final Range<PartitionKey> rangeToInclude;
+    private final int partitionTTLNumber;
+    private final PeriodDuration partitionTTL;
+    private final PartitionInfo partitionInfo;
+
+    public RangePartitionDiffer(Range<PartitionKey> rangeToInclude,
+                                int partitionTTLNumber,
+                                PeriodDuration partitionTTL,
+                                PartitionInfo partitionInfo) {
+        this.rangeToInclude = rangeToInclude;
+        this.partitionTTLNumber = partitionTTLNumber;
+        this.partitionInfo = partitionInfo;
+        this.partitionTTL = partitionTTL;
+    }
+
+    public static RangePartitionDiffer build(MaterializedView mv,
+                                             Pair<String, String> partitionRange)
```

Comment:
why not use Range<PartitionKey> as param type?

Reply:
It's not a real range. It comes from user's refresh command currently.